machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,174 @@
+
+	// spec describes the configuration of the machine os build
+	// +kubebuilder:validation:Required
+	Spec MachineOSBuildSpec `json:"spec"`
+
+	// status describes the lst observed state of this machine os build
+	// +optional
+	Status MachineOSBuildStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig string `json:"currentConfig"`
+	// desiredConfig is the desired config we want to build an image for. If currentConfig and desiredConfig are not the same, we need to build an image.
+	// +kubebuilder:validation:Required
+	DesiredConfig string `json:"desiredConfig"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Ready
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSImageReference `json:"renderedMachineOSImage"`
+	// startTime describes when this build began
+	// +kubebuilder:validation:Required
+	StartTime *metav1.Time `json:"startTime"`
+	// endTime describes when the build ended
+	// +optional
+	EndTime *metav1.Time `json:"endTime"`
+	// buildHistory contains previous iterations of failed or interrupted builds related to this one.
+	// This will keep track of failed or interrupted build names
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	BuildHistory []PriorMachineOSBuilds `json:"buildHistory,omitempty"`
```

Comment:
thought: I'm wondering if we want to somehow link the underlying build pods / ConfigMaps / etc. to this object somehow. Within BuildController, we could do that with labels and the like, but I wonder if it makes sense to link them here as well.

(I'm not asking you to add them here right now though. That could be a future addition.)

Reply:
I think I actually will add these now, if its too much, I can remove it. I will add these in my next commit